CorsixTH/Lua/app.lua

```diff
@@ -949,8 +971,11 @@ function App:writeToFileOrTmp(file, mode)
   if err then
     local tmp_file = os.tmpname()
     f = io.open(tmp_file, mode or "w")
-    self.ui:addWindow(UIInformation(self.ui,
-      { _S.errors.save_to_tmp:format(file, tmp_file, err) }))
+    if self.ui then self.ui:addWindow(UIInformation(self.ui,
+        { _S.errors.save_to_tmp:format(file, tmp_file, err) }))
+    else
+      print("A file was written to " .. tmp_file .. " because of the error: " .. err)
```

Comment:
This error is going to be pretty useless for the end user and us as written.

Maybe `"Attempt to write to " .. file .. " failed. File was written instead to temporary location " .. tmp_file .. " because of the error: " .. err`

Context errors would be even better, but they'd need to be handled outside of this function by returning the result. e.g.

```
Attempting to write configuration failed. Configuration was instead written to temporary location xxx. Your configuration changes will be lost on restart unless you copy that file to aaa
```
That is actionable and user friendly.

Reply:
The `_S` version uses a different format but generally the same:
``save_to_tmp = "The file at %s could not be used. The game has been saved to %s. Error: %s",``
E.g. when you set the config file to read-only:
![image](https://user-images.githubusercontent.com/20030128/204359361-514ccdfd-fb01-424f-a4d2-cf89d7ad79da.png)
I agree with the rewrite of the print string.

(This is more specifically about the context error part here)
I'm not sure how much I agree with end-users going around digging in temp files and moving them though. As it still may not solve the root problem (more Windows and Mac users, I imagine Linux are probably more handy but that's quite a generalisation). Plus, the temp file is saved as `.0`, I'm pretty sure we'd get people just dropping the `.0` file without changing extension unless they're told to. The current solution does include the error. So many also include the error output of why x file failed to write too some point in the future.

Reply:
In fact the `save_to_tmp` string isnt generalised enough currently.